File: src/android/app/src/main/kotlin/com/austa/superapp/features/marketplace/domain/models/Product.kt
package com.austa.superapp.features.marketplace.domain.models

import androidx.room.Embedded
import androidx.room.Entity
import androidx.room.Index
import androidx.room.TypeConverters
import kotlinx.serialization.Serializable
import com.austa.superapp.common.ProductCategory

// kotlinx.serialization v1.5.1
// androidx.room v2.5.2

/**
 * Constants for product validation
 */
private const val MAX_NAME_LENGTH = 100
private const val MAX_DESCRIPTION_LENGTH = 1000
private const val MAX_IMAGES = 10
private const val MIN_PRICE = 0.0
private const val MAX_PREREQUISITES = 5
private const val MAX_OUTCOMES = 10
private const val MAX_FORMAT_LENGTH = 50

/**
 * Data class representing detailed product information with embedded persistence support.
 * Contains validation for format, prerequisites and outcomes.
 */
@Serializable
@Embedded
data class ProductDetails(
    val duration: Int,
    val format: String,
    val prerequisites: List<String>,
    val outcomes: List<String>
) {
    init {
        require(format.length <= MAX_FORMAT_LENGTH) {
            "Format length cannot exceed $MAX_FORMAT_LENGTH characters"
        }
        require(prerequisites.size <= MAX_PREREQUISITES) {
            "Prerequisites list cannot exceed $MAX_PREREQUISITES items"
        }
        require(outcomes.size <= MAX_OUTCOMES) {
            "Outcomes list cannot exceed $MAX_OUTCOMES items"
        }
    }
}

/**
 * Data class representing a marketplace product with comprehensive validation and persistence support.
 * Implements Room Entity for local storage and Serializable for API communication.
 */
@Serializable
@Entity(
    tableName = "products",
    indices = [Index(value = ["providerId"])]
)
@TypeConverters(ProductTypeConverters::class)
data class Product(
    val id: String,
    val name: String,
    val description: String,
    val category: ProductCategory,
    val price: Double,
    val providerId: String,
    val images: List<String>,
    val details: ProductDetails,
    val createdAt: Long,
    val updatedAt: Long
) {
    init {
        require(name.length <= MAX_NAME_LENGTH) {
            "Product name cannot exceed $MAX_NAME_LENGTH characters"
        }
        require(description.length <= MAX_DESCRIPTION_LENGTH) {
            "Product description cannot exceed $MAX_DESCRIPTION_LENGTH characters"
        }
        require(price >= MIN_PRICE) {
            "Product price cannot be negative"
        }
        require(images.size <= MAX_IMAGES) {
            "Product images cannot exceed $MAX_IMAGES items"
        }
        require(id.isNotBlank()) {
            "Product ID cannot be blank"
        }
        require(providerId.isNotBlank()) {
            "Provider ID cannot be blank"
        }
        require(createdAt > 0) {
            "Created timestamp must be positive"
        }
        require(updatedAt >= createdAt) {
            "Updated timestamp must be greater than or equal to created timestamp"
        }
    }
}

File: src/web/src/lib/types/product.ts
/**
 * @fileoverview TypeScript type definitions for marketplace products in the AUSTA SuperApp platform
 * @version 1.0.0
 * @package typescript@5.0.0
 */

// Maximum value constants for validation
export const MAX_NAME_LENGTH = 100;
export const MAX_DESCRIPTION_LENGTH = 1000;
export const MAX_IMAGES = 10;
export const MIN_PRICE = 0;
export const MAX_RATING = 5;
export const MIN_RATING = 0;
export const MAX_SESSION_COUNT = 52;
export const MAX_TAGS = 20;

/**
 * Enumeration of available product categories in the marketplace
 */
export enum ProductCategory {
    DIGITAL_THERAPY = 'DIGITAL_THERAPY',
    WELLNESS_PROGRAM = 'WELLNESS_PROGRAM',
    PROVIDER_SERVICE = 'PROVIDER_SERVICE'
}

/**
 * Interface defining detailed product information
 */
export interface ProductDetails {
    /** Duration in minutes */
    duration: number;
    
    /** Delivery format (e.g., "Video", "Interactive", "Live Session") */
    format: string;
    
    /** List of prerequisites for the product */
    prerequisites: string[];
    
    /** Expected outcomes from using the product */
    outcomes: string[];
    
    /** Number of sessions included */
    sessionCount: number;
    
    /** Method of delivery (e.g., "Mobile", "Web", "In-Person") */
    deliveryMethod: string;
    
    /** Supported languages */
    languages: string[];
    
    /** Professional certifications and accreditations */
    certifications: string[];
}

/**
 * Main product interface representing items in the marketplace
 */
export interface Product {
    /** Unique identifier for the product */
    id: string;
    
    /** Product name (max 100 characters) */
    name: string;
    
    /** Detailed description (max 1000 characters) */
    description: string;
    
    /** Product category */
    category: ProductCategory;
    
    /** Price in smallest currency unit (e.g., cents) */
    price: number;
    
    /** ID of the healthcare provider offering the product */
    providerId: string;
    
    /** Array of image URLs (max 10 images) */
    images: string[];
    
    /** Detailed product information */
    details: ProductDetails;
    
    /** Creation timestamp */
    createdAt: Date;
    
    /** Last update timestamp */
    updatedAt: Date;
    
    /** Current product status */
    status: ProductStatus;
    
    /** Search and filter tags (max 20) */
    tags: string[];
    
    /** Average rating (0-5) */
    rating: number;
    
    /** Total number of reviews */
    reviewCount: number;
}

/**
 * Enumeration of product sorting options
 */
export enum ProductSortOption {
    PRICE_ASC = 'PRICE_ASC',
    PRICE_DESC = 'PRICE_DESC',
    NAME_ASC = 'NAME_ASC',
    NAME_DESC = 'NAME_DESC',
    NEWEST = 'NEWEST',
    RATING_DESC = 'RATING_DESC',
    POPULARITY = 'POPULARITY'
}

/**
 * Enumeration of possible product status values
 */
export enum ProductStatus {
    ACTIVE = 'ACTIVE',
    INACTIVE = 'INACTIVE',
    PENDING_REVIEW = 'PENDING_REVIEW',
    SUSPENDED = 'SUSPENDED'
}